qualtran/bloqs/rotations/programmable_ancilla_rotation.py

```diff
@@ -72,90 +78,96 @@ class RzViaProgrammableAncillaRotation(Bloq):
+         apply_final_correction: Whether to apply an expensive ZPow rotation at
                  the end to correct the qubit in case all measurements failed.
 
     References:
         [Simulating chemistry efficiently on fault-tolerant quantum computers](https://arxiv.org/abs/1204.0567)
         Jones et. al. 2012. Fig 4.
     """
 
-    angle: SymbolicFloat
+    exponent: SymbolicFloat
     eps: SymbolicFloat = 1e-11
-    n_rounds: SymbolicInt = 2
-    apply_final_correction: bool = False
+    n_rounds: SymbolicInt = field(default=2, kw_only=True)
+    apply_final_correction: bool = field(default=False, kw_only=True)
 
     @cached_property
     def signature(self) -> Signature:
         return Signature.build_from_dtypes(q=QBit())
 
     @classmethod
     def from_failure_probability(
-        cls, angle: SymbolicFloat, *, max_fail_probability: SymbolicFloat
-    ) -> 'RzViaProgrammableAncillaRotation':
+        cls,
+        exponent: SymbolicFloat,
+        *,
+        max_fail_probability: SymbolicFloat,
+        eps: SymbolicFloat = 1e-11,
+    ) -> 'ZPowUsingProgrammedAncilla':
         """Applies the rotation `Rz(angle)` except with some specified failure probability.
 
         Args:
-            angle: Rotation angle.
+            exponent: Rotation exponent.
             max_fail_probability: Upper bound on fail probability of the rotation gate.
+            eps: The precision of the synthesized rotation.
         """
         n_rounds = ceil(log2(1 / max_fail_probability))
-        return cls(angle, n_rounds)
+        return cls(exponent=exponent, eps=eps, n_rounds=n_rounds)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> set['BloqCountT']:
+        resources: Counter[Bloq] = Counter({CNOT(): self.n_rounds, XGate(): self.n_rounds})
+
+        n_rz = self.n_rounds + (1 if self.apply_final_correction else 0)
 
-    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         if is_symbolic(self.n_rounds):
             phi = ssa.new_symbol(r"\phi")
             eps = ssa.new_symbol(r"\epsilon")
-            resources = {(RzResourceState(phi, eps), self.n_rounds)}
+            resources[ZPowProgrammedAncilla(phi, eps)] += self.n_rounds
         else:
-            resources = {
-                (RzResourceState(2**i * self.angle, eps=self.eps / 2 ** (i + 1)), 1)
-                for i in range(int(self.n_rounds))
-            }
+            for i in range(int(self.n_rounds)):
+                resources[ZPowProgrammedAncilla(2**i * self.exponent, eps=self.eps / n_rz)] += 1
 
         if self.apply_final_correction:
-            resources |= {
-                (
-                    RzResourceState(
-                        2**self.n_rounds * self.angle, eps=self.eps / 2**self.n_rounds
-                    ),
-                    1,
-                )
-            }
+            resources[Rz(2**self.n_rounds * self.exponent, eps=self.eps / n_rz)] += 1
```

Comment:
Should this final correction be an `Rz` or a `ZPow` ?

Reply:
`ZPow`, fixed